### ISSUE TITLE
Distributed Training with 4 machines on Translation Task

### ISSUE THREAD

**Issue by 282866787** [tensorflow/tensor2tensor#944]
Hello , I want to do distributed training using four machines ,each one has 8 1080ti GPUs on En-Zh translation task, and  the t2t-version is 1.6.5. I have seen the other similar issues , and 
 the distributed_training.md ,but  I still have some confusion. If the four machines named M1, M2 ,M3 ,M4 , the first step is to make the TF-Config. What I want is to make 4 machines to train together , thus I am not sure how to define the master and ps. And the second question is if four machines can do synchronous training ?
Hope someone can give me some advice, thanks

**Comment by 282866787** [comment]
Update:

This is my ps :
![image](https://user-images.githubusercontent.com/26756710/42867438-3309c8ca-8aa2-11e8-9daa-e7af3d5abd3d.png)
This is my worker1:
![image](https://user-images.githubusercontent.com/26756710/42867454-3e9ae58e-8aa2-11e8-8932-d33b15ab91e0.png)
This is my worker2:
![image](https://user-images.githubusercontent.com/26756710/42867472-46a0cee2-8aa2-11e8-8e8d-7aceb5e60d35.png)

Now it seems like the worker1 hang out and the worker2 is still waiting for worker1 to creat the session .
I use the t2t-make-configs to produce the TF_CONFIG , does anyone can show me some advice ,thanks !

**Comment by 282866787** [comment]
Now I use one single gpu to test if the distributed training can run normarly, gpu0 is the ps ,and worker1 is the gpu1 ,worker2 is the gpu2

**Comment by 282866787** [comment]
@rsepassi I have tired sucessfully with on the single machine , 1 ps and two masters  in asynchronous way. But I really want to train models in synchronous , and the code "--sync" if len(masters) == 1 else "", make me confused , is there any solution ? I'm looking forward to your advising, appreciate you.

**Comment by martinpopel** [comment]
I have never tried distributed training, but somewhere here in old GitHub issues someone wrote that the naming "parameter server" and "master" is confusing for newcomers (and maybe should be switched). Maybe you want one master and two parameter servers? Also, I thing the synchronous training is the default (as it is reported to have better results than the asynchronous training).

**Comment by 282866787** [comment]
@martinpopel yeah , thanks for your response ,and I have read the issues before and I successfully trained in asynchronous way. The ps is to compute the parameter gradients and the master (worker ) is to do the training job(forward training and send parameter to ps). But masters more than 1 can't use sync mode , I believe there may be something wrong I don't know

**Comment by 282866787** [comment]
@lukaszkaiser @rsepassi Could you please show me some advice about how to define the masters and ps in T2T-distributed training task . Because of the setting , --sync can only support with the single master, I just want to solve 4 machines to synchronous train.

**Comment by 282866787** [comment]
I have solve the problem now , but I meet a new one ,when I use multiple machines to train , the master got an error like this 
![image](https://user-images.githubusercontent.com/26756710/42983963-65bdbdd2-8c1c-11e8-88de-35409d33a39f.png)
CUDA_VISIBLE_DEVICES=7 TF_CONFIG='{"cluster": {"master": ["192.168.0.217:10000"], "ps": ["192.168.0.217:5000", "192.168.0.230:5000"]}, "environment": "cloud", "task": {"index": 0, "type": "master"}}' PYTHONPATH=./ python3 tensor2tensor/bin/t2t-trainer --data_dir=/home/nfs_share/train  --problems=wmt_zhen_tokens_32k --model=transformer --hparams_set=transformer_base --output_dir=/home/nfs_share/model --master=grpc://192.168.0.217:10000 --ps_replicas=2 --worker_replicas=1 --worker_gpu=1 --worker_id=0 --worker_job='/job:master' --ps_gpu=6 --schedule=train --sync
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5 TF_CONFIG='{"cluster": {"master": ["192.168.0.217:10000"], "ps": ["192.168.0.217:5000", "192.168.0.230:5000"]}, "environment": "cloud", "task": {"index": 0, "type": "ps"}}' PYTHONPATH=./ python3 tensor2tensor/bin/t2t-trainer --data_dir=/home/nfs_share/train  --problems=wmt_zhen_tokens_32k --model=transformer --hparams_set=transformer_base --output_dir=/home/nfs_share/model --master=grpc://192.168.0.217:5000 --schedule=run_std_server
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5 TF_CONFIG='{"cluster": {"master": ["192.168.0.217:10000"], "ps": ["192.168.0.217:5000", "192.168.0.230:5000"]}, "environment": "cloud", "task": {"index": 1, "type": "ps"}}' PYTHONPATH=./ python3 tensor2tensor/bin/t2t-trainer --data_dir=/home/client_nfs_share/train  --problems=wmt_zhen_tokens_32k --model=transformer --hparams_set=transformer_base --output_dir=/home/client_nfs_share/model --master=grpc://192.168.0.230:5000 --schedule=run_std_server
The config is above , is there any mistake ?

**Comment by 282866787** [comment]
@martinpopel Hi,I have tried several methods to build my distributed training system , but the training speed is fallacious. As I want, several machines with one master and muti ps(3 ps),the training speed depends on the slowest one of the ps. But what I got it is almost 200s every 100steps in En-De training, I'm very confused about this. What's more ,the t2t defines the master and ps , but in conventional tensorflow distributed training tutorial, there should be worker and ps. The ps plays the role of updating variables like master in T2T , and the worker computes the gradients  parallel like ps in T2T. This really makes me amazing! I have ask questions in gitter ,but no one answer me . I hope you can give me some advice !

**Comment by JohannesTK** [comment]
You can have a look at the updated distributed training docs: https://github.com/tensorflow/tensor2tensor/blob/master/docs/distributed_training.md

**Comment by rsepassi** [comment]
And make sure the Tensorflow version is the same on all machines.
